servers/physics/space_sw.cpp

```diff
@@ -37,7 +37,7 @@ _FORCE_INLINE_ static bool _match_object_type_query(CollisionObjectSW *p_object,
 	if (p_object->get_type()==CollisionObjectSW::TYPE_AREA)
 		return p_type_mask&PhysicsDirectSpaceState::TYPE_MASK_AREA;
 
-	if ((p_object->get_layer_mask()&p_layer_mask)==0)
+	if (((p_object->get_layer_mask()&p_layer_mask)==0) && ((p_object->get_collision_mask()&p_layer_mask)==0))
```

Comment:
That's incorrect, it should be:
`(A.layer & B.collision) || (A.collision & B.layer)`,
while you have:
`(A.layer & `**`B.layer`**`) || (A.collision & B.layer)`